axum-extra/src/extract/or_default.rs

```diff
@@ -0,0 +1,175 @@
+use axum::{
+    async_trait,
+    extract::{FromRequest, RequestParts},
+};
+use std::{
+    convert::Infallible,
+    ops::{Deref, DerefMut},
+};
+
+/// Use an extractor's default value if it fails.
+///
+/// `OrDefault` wraps another extractors, that implements [`Default`], and uses its default value
+/// if the inner extractor fails.
+///
+/// # Example
+///
+/// Serializing query params for pagination and using default values for missing params:
+///
+/// ```rust
+/// use axum_extra::extract::OrDefault;
+/// use axum::extract::Query;
+/// use serde::Deserialize;
+///
+/// #[derive(Default, Deserialize)]
+/// struct Pagination {
+///     #[serde(default)]
+///     page: Page,
+///     #[serde(default)]
+///     per_page: PerPage,
+/// }
+///
+/// #[derive(Deserialize, Copy, Clone)]
+/// #[serde(transparent)]
+/// struct Page(usize);
+///
+/// impl Default for Page {
+///     fn default() -> Self {
+///         Self(1)
+///     }
+/// }
+///
+/// #[derive(Deserialize, Copy, Clone)]
+/// #[serde(transparent)]
+/// struct PerPage(usize);
+///
+/// impl Default for PerPage {
+///     fn default() -> Self {
+///         Self(30)
+///     }
+/// }
+///
+/// async fn handler(
+///     pagination: OrDefault<Query<Pagination>>,
```

Comment:
Why is `OrDefault` even needed in this case? If there is not `?query_params`, does `Query` fail rather than passing an empty string to `serde_urlencoded`? I think the latter is what actix-web does and should make this example work without `OrDefault`.

Reply:
Ah yeah thats a good point. If there is no query string it gives an empty string to `serde_urlencoded` so that would work without `OrDefault`. If some params are present but invalid `OrDefault` would catch the error and return the default instead. I'll update the example to mention that.

Reply:
Hm maybe pagination isn't the best example 🤔